src/algebra/group/defs.lean

```diff
@@ -174,14 +174,32 @@ theorem mul_left_inj (a : G) {b c : G} : b * a = c * a ↔ b = c :=
 
 end right_cancel_semigroup
 
+/-- Typeclass for expressing that a type `M` with multiplication and a one satisfies
+`1 * a = a` and `a * 1 = a` for all `a : M`. -/
+@[ancestor has_one has_mul]
+class mul_one_class (M : Type u) extends has_one M, has_mul M :=
+(one_mul : ∀ (a : M), 1 * a = a)
+(mul_one : ∀ (a : M), a * 1 = a)
+
+/-- Typeclass for expressing that a type `M` with addition and a zero satisfies
+`0 + a = a` and `a + 0 = a` for all `a : M`. -/
+@[ancestor has_zero has_add]
+class add_zero_class (M : Type u) extends has_zero M, has_add M :=
+(zero_add : ∀ (a : M), 0 + a = a)
+(add_zero : ∀ (a : M), a + 0 = a)
+
+attribute [to_additive] mul_one_class
+
 /-- A `monoid` is a `semigroup` with an element `1` such that `1 * a = a * 1 = a`. -/
 @[ancestor semigroup has_one]
 class monoid (M : Type u) extends semigroup M, has_one M :=
 (one_mul : ∀ a : M, 1 * a = a) (mul_one : ∀ a : M, a * 1 = a)
+
 /-- An `add_monoid` is an `add_semigroup` with an element `0` such that `0 + a = a + 0 = a`. -/
 @[ancestor add_semigroup has_zero]
 class add_monoid (M : Type u) extends add_semigroup M, has_zero M :=
 (zero_add : ∀ a : M, 0 + a = a) (add_zero : ∀ a : M, a + 0 = a)
+
```

Comment:
Similarlyto the comment below (https://github.com/leanprover-community/mathlib/pull/6786/files#r598136450), the instances are missing that go from `monoid` to `mul_one_class`. I remember you tried in the last PR by adjusting the `extends` clauses.

Can you try again, this time adding the instances manually?

Reply:
I've just added https://github.com/leanprover-community/mathlib/pull/6786/commits/f867c47f6eb458bb9fcfdf30c6941fb14a2a77c5 which is one last attempt to do this using `extends`.

I'll add the instances manually if that effort is unsuccessful (e.g., if I give up in frustration at the painfully-slow debug cycle again).

Reply:
I can't believe it. After just one more fix, it now builds 🎉